sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala

```diff
@@ -45,8 +56,160 @@ class SidechainNodeViewSynchronizer(networkControllerRef: ActorRef,
+            penalizeMisbehavingPeer(remote)
+            log.warn(s"Received additional bytes after block. Declared id ${encoder.encodeId(id)} from ${remote.toString}")
+          }
+          Some(mod)
+        case _ =>
+          // Penalize peer and do nothing - it will be switched to correct state on CheckDelivery
+          penalizeMisbehavingPeer(remote)
+          log.warn(s"Failed to parse modifier with declared id ${encoder.encodeId(id)} from ${remote.toString}")
+          None
+      }
+    }
+  }
+
+  /**
+   * Method copied from the parent class, with the addition of isReindexing check: we must avoid to penalize
+   * a peer if receiving a block during reindex
+   *
+   * Move `pmod` to `Invalid` if it is permanently invalid, to `Received` otherwise
+   */
+  @SuppressWarnings(Array("org.wartremover.warts.IsInstanceOf"))
+  private def validateAndSetStatus(remote: ConnectedPeer, pmod: SidechainBlock): Boolean = {
+    if (isReindexing) {
+      log.error("Got modifier while reindexing - will be discarded")
+      deliveryTracker.setReceived(pmod.id, remote)
+      false
+    }else{
+      historyReaderOpt match {
+        case Some(hr) =>
+          hr.applicableTry(pmod) match {
+            case Failure(e) if e.isInstanceOf[MalformedModifierError] =>
+              log.warn(s"Modifier ${pmod.encodedId} is permanently invalid", e)
+              deliveryTracker.setInvalid(pmod.id)
+              penalizeMisbehavingPeer(remote)
+              false
+            case _ =>
+              deliveryTracker.setReceived(pmod.id, remote)
+              true
+          }
+        case None =>
+          log.error("Got modifier while history reader is not ready")
+          deliveryTracker.setReceived(pmod.id, remote)
+          true
+      }
+    }
+  }
+
+  /**
+   * Method copied as-is  from the parent class to be used by the previous
+   * TODO: this method can be removed once inherited method from Sparkz will be changed to protected visibility
+   *
+   * Get modifiers from remote peer,
+   * filter out spam modifiers and penalize peer for spam
+   *
+   * @return ids and bytes of modifiers that were requested by our node
+   */
+  private def processSpam(remote: ConnectedPeer,
+                          typeId: ModifierTypeId,
+                          modifiers: Map[ModifierId, Array[Byte]]): Map[ModifierId, Array[Byte]] = {
+
+    val (requested, spam) = modifiers.partition { case (id, _) =>
+      deliveryTracker.status(id) == Requested
+    }
+
+    if (spam.nonEmpty) {
+      log.info(s"Spam attempt: peer $remote has sent a non-requested modifiers of type $typeId with ids" +
+        s": ${spam.keys.map(encoder.encodeId)}")
+      penalizeSpammingPeer(remote)
+    }
+    requested
+  }
+
+
+  protected def changedHistoryEvent: Receive = {
+    case ChangedHistory(sHistory: SidechainHistory) =>
+      historyReaderOpt = Some(sHistory)
+      if (!historyReaderOpt.isEmpty){
+        isReindexing = sHistory.isReindexing()
+      }
```

Comment:
```suggestion
      historyReaderOpt = Some(sHistory)
      isReindexing = sHistory.isReindexing()
```
we are sure that sHistory is present